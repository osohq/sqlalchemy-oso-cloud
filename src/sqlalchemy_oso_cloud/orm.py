"""
Utilities for [declaratively mapping](https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#orm-declarative-mapping)
[authorization data](https://www.osohq.com/docs/authorization-data) in your ORM models.
"""

from sqlalchemy.orm import MappedColumn, mapped_column, relationship

class Resource:
  """
  A mixin to indicate that an ORM model corresponds to an Oso resource.
  """
  pass

_RELATION_INFO_KEY = "_oso.relation"
_ATTRIBUTE_INFO_KEY = "_oso.attribute"

# TODO: should this be on columns or relationships?
# TODO: types
def relation(*args, **kwargs):
  """
  A wrapper around [`sqlalchemy.orm.relationship`](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship)
  that indicates that the relationship corresponds to `has_relation` facts in Oso with the following three arguments:
  1. the resource this relationship is declared on,
  2. the name of this relationship, and
  3. the resource that the relationship points to.

  Accepts all of the same arguments as [`sqlalchemy.orm.relationship`](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship).
  """
  rel = relationship(*args, **kwargs)
  rel.info[_RELATION_INFO_KEY] = None
  return rel

def attribute(*args, **kwargs) -> MappedColumn:
  """
  A wrapper around [`sqlalchemy.orm.mapped_column`](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column)
  that indicates that the attribute corresponds to `has_{attribute_name}` facts in Oso with the following two arguments:
  1. the resource this attribute is declared on, and
  2. the attribute value

  Accepts all of the same arguments as [`sqlalchemy.orm.mapped_column`](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column).
  """
  col = mapped_column(*args, **kwargs)
  col.column.info[_ATTRIBUTE_INFO_KEY] = None
  return col