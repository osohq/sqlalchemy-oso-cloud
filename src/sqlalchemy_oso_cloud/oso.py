import os
import yaml

from typing import TypedDict
from oso_cloud import Oso
from sqlalchemy import select
from sqlalchemy.orm import Mapper, registry, ColumnProperty, Relationship
from sqlalchemy.sql.elements import NamedColumn
from tempfile import NamedTemporaryFile

from .orm import ATTRIBUTE_INFO_KEY, Resource, RELATION_INFO_KEY

class FactConfig(TypedDict):
  query: str

class LocalAuthorizationConfig(TypedDict):
  facts: dict[str, FactConfig]
  sql_types: dict[str, str]

def generate_local_authorization_config(registry: registry) -> LocalAuthorizationConfig:
  facts: dict[str, FactConfig] = {}
  sql_types: dict[str, str] = {}

  for mapper in registry.mappers:
    if not issubclass(mapper.class_, Resource):
      continue
    id = mapper.get_property("id")
    if not isinstance(id, ColumnProperty):
      raise ValueError("Oso id must be a column")
    if len(id.columns) != 1:
      raise ValueError("Oso id must be a single column")
    id_column = id.columns[0]
    sql_types[mapper.class_.__name__] = str(id_column.type)
    for attr in mapper.attrs:
      if isinstance(attr, Relationship) and RELATION_INFO_KEY in attr.info:
        key, query = gen_relation_binding(attr, mapper, id_column)
        facts[key] = query
      elif isinstance(attr, ColumnProperty) and ATTRIBUTE_INFO_KEY in attr.columns[0].info:
        key, query = gen_attribute_binding(attr, mapper, id_column)
        sql_types[attr.key.capitalize()] = str(attr.columns[0].type)
        facts[key] = query

  return {
    "facts": facts,
    "sql_types": sql_types,
  }

def gen_relation_binding(relationship: Relationship, mapper: Mapper, id_column: NamedColumn) -> tuple[str, FactConfig]:
  if len(relationship.local_columns) != 1:
    raise ValueError(f"Oso relation {relationship.key} must be to a single column")
  local_column = list(relationship.local_columns)[0]
  key = f"has_relation({mapper.class_.__name__}:_, {relationship.key}, {relationship.entity.class_.__name__}:_)"
  query = select(id_column, local_column)
  return key, {
    "query": str(query),
  }

def gen_attribute_binding(attribute: ColumnProperty, mapper: Mapper, id_column: NamedColumn) -> tuple[str, FactConfig]:
  if len(attribute.columns) != 1:
    raise ValueError(f"Oso attribute {attribute.key} must be a single column")
  column = attribute.columns[0]
  key = f"has_{attribute.key}({mapper.class_.__name__}:_, {attribute.key.capitalize()}:_)"
  query = select(id_column, column)
  return key, {
    "query": str(query),
  }


# TODO: what if they want multiple DBs/registries?
oso: Oso | None = None

def init(registry: registry, **kwargs):
  kwargs = { **kwargs }
  if "url" not in kwargs:
    kwargs["url"] = os.getenv("OSO_URL", "https://api.osohq.com")
  if "api_key" not in kwargs:
    kwargs["api_key"] = os.getenv("OSO_API_KEY")
  if "data_bindings" in kwargs:
    # just need to conditionally close/delete the temporary file if it was created
    raise NotImplementedError("manual data_bindings are not supported yet")
  with NamedTemporaryFile(mode="w") as f:
    config = generate_local_authorization_config(registry)
    yaml.dump(config, f)
    f.flush()
    kwargs["data_bindings"] = f.name
    global oso
    oso = Oso(**kwargs)
  
def get_oso() -> Oso:
  if oso is None:
    raise RuntimeError("sqlalchemy_oso_cloud must be initialized before getting the Oso client")
  return oso
