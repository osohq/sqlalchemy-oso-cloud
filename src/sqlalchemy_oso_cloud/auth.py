from sqlalchemy.orm import with_loader_criteria
from sqlalchemy import literal_column, ColumnClause
from oso_cloud import Value
from typing import Set, Type, Callable
from .orm import Resource
from .oso import get_oso

__all__ = ['authorized', '_apply_authorization_options']


def extract_unique_models(column_descriptions) -> Set[Type]:
    """Extract all models being queried from column descriptions"""
    models = set()
    
    for desc in column_descriptions:
        if desc['entity'] is not None:
            models.add(desc['entity'])
    return models


def create_auth_criteria_for_model(model: Type, actor: Value, action: str) -> Callable:
    """Create authorization criteria for a specific model"""
    oso = get_oso()
    
    sql_filter = oso.list_local(
        actor=actor,
        action=action,
        resource_type=model.__name__,
        column=f"{model.__tablename__}.id"
    )
    
    criteria: ColumnClause = literal_column(sql_filter)
    return lambda cls: criteria


def authorized(actor: Value, action: str, *models: Type) -> list:
    """
    Create authorization options for use with .options()
    
    This function can be used with both Select and Query objects from the SQLAlchemy library:
    
    Examples:
        # With standard SQLAlchemy select
        from sqlalchemy import select
        from sqlalchemy_oso_cloud import authorized

        stmt = select(Document).options(*authorized(user, "read", Document))
        docs = session.execute(stmt).scalars().all()

        # With Query
        docs = session.query(Document).options(*authorized(user, "read", Document)).all()

    :param actor: The actor performing the action
    :param action: The action the actor is performing  
    :param models: The model classes to authorize against
    :return: List of loader criteria options for use with .options()
    """
    
    if len(models) > 1:
        raise ValueError("Currently only single model authorization is supported.")
    if len(models) == 0:
        raise ValueError("Must provide a model to authorize against.")

    options = []
    for model in models:
        if not issubclass(model, Resource):
            continue
        auth_criteria = create_auth_criteria_for_model(model, actor, action)
        options.append(
            with_loader_criteria(
                model,
                auth_criteria,
                include_aliases=True
            )
        )
    return options


def _apply_authorization_options(query_obj, actor: Value, action: str):
    """
    Apply authorization to any query-like object that has column_descriptions and options()
    
    This works with both Select and Query objects.
    """
    models = extract_unique_models(query_obj.column_descriptions)
    auth_options = authorized(actor, action, *models)
    return query_obj.options(*auth_options)
