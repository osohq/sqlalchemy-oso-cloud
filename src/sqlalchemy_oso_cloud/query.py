import sqlalchemy.orm
from oso_cloud import Value
from typing import TypeVar

from .auth import _apply_authorization_options
from .orm import Resource
from .oso import get_oso

T = TypeVar("T")
Self = TypeVar("Self", bound="Query")

# todo - multiple permissions for multiple main models
class Query(sqlalchemy.orm.Query[T]):
  def __init__(self, *args, **kwargs):
      super().__init__(*args, **kwargs)
      self.oso = get_oso()
      self.filter_cache = {}

  def authorized(self, actor: Value, action: str) -> Self:
    return _apply_authorization_options(self, actor, action)