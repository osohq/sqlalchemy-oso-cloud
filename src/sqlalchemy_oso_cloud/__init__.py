from .session import Session
from .query import Query
from .oso import init, get_oso
from .select import AuthorizedSelect, select

__all__ = ["Session", "Query", "init", "get_oso", "select", "AuthorizedSelect"]