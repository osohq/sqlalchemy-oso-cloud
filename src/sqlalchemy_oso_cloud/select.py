from sqlalchemy.sql import Select
from sqlalchemy import select as sqlalchemy_select, literal_column
from sqlalchemy.orm import DeclarativeBase, with_loader_criteria
from oso_cloud import Value
from typing import  Self, Type
from .oso import get_oso

class AuthorizedSelect(Select):
    """A Select subclass that adds authorization functionality"""

    inherit_cache = True
    
    def __init__(self, *args, **kwargs):
        # Handle case where we're wrapping an existing Select e.g. AuthorizedSelect(existing_Select)
        if len(args) == 1 and isinstance(args[0], Select):
            select_stmt = args[0]
            super().__init__()
            self.__dict__.update(select_stmt.__dict__)
        else:
            super().__init__(*args, **kwargs)
        
        self._oso = get_oso()
        self._auth_cache = {}
    
    def authorized_for(self, actor: Value, action: str) -> Self:
        """Add authorization filtering to the select statement"""
        models = self._extract_unique_models()
        options = []
        
        for model in models:
            auth_criteria = self._create_auth_filter(model, actor, action)
            options.append(
                with_loader_criteria(
                    model,
                    auth_criteria,
                    include_aliases=True
                )
            )
        return self.options(*options)
    
    def _extract_unique_models(self):
        """Extract all models being queried"""
        models = set()
        
        for desc in self.column_descriptions:
            if desc['entity'] is not None:
                models.add(desc['entity'])
        return models

    def _create_auth_filter(self, model: Type[DeclarativeBase], actor: Value, action: str):
        """Create authorization filter for a model"""
        
        sql_filter = self._oso.list_local(
            actor=actor,
            action=action,
            resource_type=model.__name__,
            column=f"{model.__tablename__}.id"
        )
        
        criteria = literal_column(sql_filter)

        return lambda cls: criteria

def select(*args, **kwargs) -> AuthorizedSelect:
    """
    Create an AuthorizedSelect statement
    
    This is a drop-in replacement for sqlalchemy.select() that adds
    authorization capabilities via the .authorized() method.
    
    Example:
        from sqlalchemy_oso_cloud import select

        stmt = select(Document).where(Document.private == True).authorized(actor, "read")
        documents = session.execute(stmt)
    """
    return AuthorizedSelect(*args, **kwargs)