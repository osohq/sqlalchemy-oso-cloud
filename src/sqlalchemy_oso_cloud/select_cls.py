from sqlalchemy.sql import Select
from sqlalchemy import literal_column, ColumnClause
from sqlalchemy.orm import DeclarativeBase, with_loader_criteria
from oso_cloud import Value
from typing import  TypeVar, Type
from .oso import get_oso
from .auth import _apply_authorization_options

Self = TypeVar("Self", bound="AuthorizedSelect")

class AuthorizedSelect(Select):
    """A Select subclass that adds authorization functionality"""

    inherit_cache = True
    
    def __init__(self, *args, **kwargs):
        # Handle case where we're wrapping an existing Select e.g. AuthorizedSelect(existing_Select)
        if len(args) == 1 and isinstance(args[0], Select):
            select_stmt = args[0]
            super().__init__()
            self.__dict__.update(select_stmt.__dict__)
        else:
            super().__init__(*args, **kwargs)
        
        self._oso = get_oso()
    
    def authorized(self: Self, actor: Value, action: str) -> Self:
        """Add authorization filtering to the select statement"""
        return _apply_authorization_options(self, actor, action)
    
    
def select(*args, **kwargs) -> AuthorizedSelect:
    """
    Create an AuthorizedSelect statement
    
    This is a drop-in replacement for sqlalchemy.select() that adds
    authorization capabilities via the .authorized() method.
    
    Example:
        from sqlalchemy_oso_cloud import select

        stmt = select(Document).where(Document.private == True).authorized(actor, "read")
        documents = session.execute(stmt)
    """
    return AuthorizedSelect(*args, **kwargs)
