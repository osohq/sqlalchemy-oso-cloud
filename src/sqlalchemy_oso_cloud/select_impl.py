import sqlalchemy.sql
from sqlalchemy.orm.attributes import InstrumentedAttribute
from oso_cloud import Value
from typing import  TypeVar, Generic, Any, overload, Type, Tuple, Union
from .auth import _apply_authorization_options

Self = TypeVar("Self", bound="Select")
_T = TypeVar("_T", bound=Tuple[Any, ...])
T = TypeVar("T")
T1 = TypeVar("T1")
T2 = TypeVar("T2")
T3 = TypeVar("T3")
T4 = TypeVar("T4")

class Select(sqlalchemy.sql.Select[_T], Generic[_T]):
    """A Select subclass that adds authorization functionality"""

    inherit_cache = True
    """Internal SQLAlchemy caching optimization"""
    
    
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
    
    def authorized(self: Self, actor: Value, action: str) -> Self:
        """Add authorization filtering to the select statement"""
        return _apply_authorization_options(self, actor, action)

# Single entity overload
@overload
def select(entity: Type[T]) -> Select[Tuple[T]]: ...

# Single column overload
@overload
def select(column: InstrumentedAttribute[T]) -> Select[Tuple[T]]: ...

@overload
def select(
    arg1: Union[Type[T1], InstrumentedAttribute[T1]], 
    arg2: Union[Type[T2], InstrumentedAttribute[T2]]) -> Select[Tuple[T1, T2]]: ...

@overload
def select(arg1: Union[Type[T1], InstrumentedAttribute[T1]], 
           arg2: Union[Type[T2], InstrumentedAttribute[T2]], 
           arg3: Union[Type[T3], InstrumentedAttribute[T3]]) -> Select[Tuple[T1, T2, T3]]: ...

@overload
def select(arg1: Union[Type[T1], InstrumentedAttribute[T1]], 
           arg2: Union[Type[T2], InstrumentedAttribute[T2]], 
           arg3: Union[Type[T3], InstrumentedAttribute[T3]], 
           arg4: Union[Type[T4], InstrumentedAttribute[T4]]) -> Select[Tuple[T1, T2, T3, T4]]: ...

@overload
def select(*entities: Any) -> Select[Any]: ...    
def select(*entities, **kwargs) -> Select[Any]:
    """
    Create an sqlalchemy_oso_cloud.Select() object

    This is a drop-in replacement for sqlalchemy.select() that adds
    authorization capabilities via the .authorized() method.
    
    Example:
        from sqlalchemy_oso_cloud import select

        stmt = select(Document).where(Document.private == True).authorized(actor, "read")
        documents = session.execute(stmt)
    """
    return Select(*entities, **kwargs)
