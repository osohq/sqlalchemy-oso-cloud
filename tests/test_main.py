from oso_cloud import Oso, Value
from sqlalchemy import text
from sqlalchemy.orm import Session
from .models import Document

import sqlalchemy_oso_cloud

# This is the part our goal is to make nicer
def test_manual(oso: Oso, session: Session, alice: Value, bob: Value):
  filter = oso.list_local(alice, "read", "Document", "document.id")
  documents = session.query(Document).filter(text(filter)).all()
  assert len(documents) == 1
  assert documents[0].id == 1
  filter = oso.list_local(bob, "read", "Document", "document.id")
  documents = session.query(Document).filter(text(filter)).all()
  assert len(documents) == 1
  assert documents[0].id == 2

def test_library(oso_session: sqlalchemy_oso_cloud.Session, alice: Value):
  documents = oso_session.query(Document).authorized(alice, "read").all()
  assert len(documents) == 1
  assert documents[0].id == 1

def test_bob_can_read_different_document(oso_session: sqlalchemy_oso_cloud.Session, bob: Value):
  documents = oso_session.query(Document).authorized(bob, "read").all()
  assert len(documents) == 1
  assert documents[0].id == 2

def test_bob_cannot_eat_documents(oso_session: sqlalchemy_oso_cloud.Session, bob: Value):
  documents = oso_session.query(Document).authorized(bob, "eat").all()
  assert len(documents) == 0

def test_authorize_with_filter(oso_session: sqlalchemy_oso_cloud.Session, alice: Value):
  documents = (
      oso_session.query(Document)
      .filter(Document.id == 1)
      .authorized(alice, "read")
      .filter(Document.content == "hello")
      .all()
  )
  assert len(documents) == 1
  assert documents[0].id == 1

def test_authorize_doesnt_bring_in_filtered(oso_session: sqlalchemy_oso_cloud.Session, alice: Value):
  documents = (
      oso_session.query(Document)
      .filter(Document.content == "world")
      .authorized(alice, "read")
      .all()
  )
  assert len(documents) == 0
