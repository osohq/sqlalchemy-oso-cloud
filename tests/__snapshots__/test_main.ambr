# serializer version: 1
# name: test_local_authorization_config_snapshot
  '''
  facts:
    has_relation(Document:_, organization, Organization:_):
      query: "SELECT document.id, organization.id AS id_1 \nFROM document, organization\
        \ \nWHERE organization.id = document.organization_id"
    has_relation(Document:_, team, Team:_):
      query: "SELECT document.id, document.team_id \nFROM document"
    has_relation(Organization:_, documents, Document:_):
      query: "SELECT organization.id, document.id AS id_1 \nFROM organization, document\
        \ \nWHERE organization.id = document.organization_id"
    has_status(Document:_, String:_):
      query: "SELECT document.id, document.status \nFROM document"
    is_public(Document:_):
      query: "SELECT document.id \nFROM document \nWHERE document.is_public"
  sql_types:
    Document: INTEGER
    Organization: INTEGER
    Team: INTEGER
  
  '''
# ---
